### PR TITLE
hir::Expr

### PR DESCRIPTION
**Pull request by flodiebold**
Still a bit to do, but I already adapted `FnScopes` and thought I'd get feedback already.

### CODE REVIEW

File: crates/ra_hir/src/expr.rs [hunk 1]
```diff
@@ -0,0 +1,740 @@
...
+    pat_syntax_mapping: FxHashMap<LocalSyntaxPtr, PatId>,
+    pat_syntax_mapping_back: FxHashMap<PatId, LocalSyntaxPtr>,
+}
+
+impl Body {
+    pub fn expr(&self, expr: ExprId) -> &Expr {
+        &self.exprs[expr]
+    }
+
+    pub fn pat(&self, pat: PatId) -> &Pat {
```
Comment (matklad) [inline crates/ra_hir/src/expr.rs]: These could be Index impls

File: crates/ra_hir/src/expr.rs [hunk 2]
```diff
@@ -0,0 +1,740 @@
...
+/// An item body together with the mapping from syntax nodes to HIR expression
+/// IDs. This is needed to go from e.g. a position in a file to the HIR
+/// expression containing it; but for type inference etc., we want to operate on
+/// a structure that is agnostic to the actual positions of expressions in the
+/// file, so that we don't recompute the type inference whenever some whitespace
+/// is typed.
+#[derive(Debug, Eq, PartialEq)]
+pub struct BodySyntaxMapping {
+    body: Arc<Body>,
+    expr_syntax_mapping: FxHashMap<LocalSyntaxPtr, ExprId>,
```
Comment (matklad) [inline crates/ra_hir/src/expr.rs]: Would it be pemature to extract IdBiMap here?

File: crates/ra_hir/src/expr.rs [hunk 3]
```diff
@@ -0,0 +1,740 @@
...
+            Expr::Field { expr, .. }
+            | Expr::Try { expr }
+            | Expr::Cast { expr, .. }
+            | Expr::Ref { expr, .. }
+            | Expr::UnaryOp { expr, .. } => {
+                f(*expr);
+            }
+        }
+    }
+}
```
Comment (matklad) [inline crates/ra_hir/src/expr.rs]: It probably is’t worth it, but we can try to generate this from grammar.ron

File: crates/ra_hir/src/function/scope.rs [hunk 4]
```diff
@@ -34,55 +36,132 @@ pub struct ScopeData {
...
+        context_expr: ExprId,
+        name: Name,
+    ) -> Option<&'a ScopeEntry> {
+        let mut shadowed = FxHashSet::default();
+        let ret = self
+            .scope_chain_for(context_expr)
+            .flat_map(|scope| self.entries(scope).iter())
+            .filter(|entry| shadowed.insert(entry.name()))
+            .filter(|entry| entry.name() == &name)
+            .nth(0);
```
Comment (matklad) [inline crates/ra_hir/src/function/scope.rs]: .find?

File: crates/ra_hir/src/function/scope.rs [hunk 5]
```diff
@@ -34,55 +36,132 @@ pub struct ScopeData {
...
+                pat,
+            }),
+            p => p.walk_child_pats(|pat| self.add_bindings(body, scope, pat)),
+        }
+    }
+    fn add_params_bindings(&mut self, scope: ScopeId, params: &[PatId]) {
+        let body = Arc::clone(&self.body);
+        params
+            .into_iter()
+            .for_each(|it| self.add_bindings(&body, scope, *it));
```
Comment (matklad) [inline crates/ra_hir/src/function/scope.rs]: btw, using ‘it’ for “don’t care” bindings is borrowed from Kotlin(which got it from groovy) Not sure it’s a good idea 🤷‍♂️. I use it when I am lazy about inventing a good short name, but at least I am explicit about the fact that the name is bad :-)

File: crates/ra_hir/src/function/scope.rs [hunk 6]
```diff
@@ -127,203 +211,95 @@ impl FnScopes {
...
-    }
-    fn set_scope(&mut self, node: SyntaxNodeRef, scope: ScopeId) {
-        self.scope_for.insert(LocalSyntaxPtr::new(node), scope);
-    }
     fn scope_for(&self, node: SyntaxNodeRef) -> Option<ScopeId> {
         node.ancestors()
             .map(LocalSyntaxPtr::new)
-            .filter_map(|it| self.scope_for.get(&it).map(|&scope| scope))
+            .filter_map(|ptr| self.syntax_mapping.syntax_expr(ptr))
+            .filter_map(|it| self.scopes.scope_for(it))
```
Comment (matklad) [inline crates/ra_hir/src/function/scope.rs]: find_map

File: crates/ra_hir/src/source_binder.rs [hunk 7]
```diff
@@ -87,6 +87,20 @@ fn module_from_source(
     Ok(Some(Module::new(db, source_root_id, module_id)?))
 }
 
+pub fn function_from_position(
+    db: &impl HirDatabase,
+    position: FilePosition,
+) -> Cancelable<Option<Function>> {
+    let file = db.source_file(position.file_id);
+    let fn_def = if let Some(f) = find_node_at_offset::<ast::FnDef>(file.syntax(), position.offset)
```
Comment (matklad) [inline crates/ra_hir/src/source_binder.rs]: crtry works here?

**Comment by matklad** [comment]
Overall approach looks good to me!

File: crates/ra_hir/src/expr.rs (hunk 2 above)
Comment (flodiebold) [reply crates/ra_hir/src/expr.rs]: It at least makes the naming easier :smile:

Comment (flodiebold) [reply crates/ra_hir/src/expr.rs]: Hm, there are a few cases where the mapping isn't perfectly bidirectional however -- e.g. with a `ParenExpr`, both it and the inner expr point to the same HIR expr. Which may not be a great thing either...

Comment (matklad) [reply crates/ra_hir/src/expr.rs]: Hm, makes sense. That's more arguments in favor of encapsulating this fuzziness into a separate type.

**Comment by matklad** [comment]
Given that I am refactoring API around defs, I wonder if it makes sense to try to land this piece wise?

**Comment by flodiebold** [comment]
@matklad Sure, I just pushed the current working state, I implemented your smaller comments and added a query for function signatures.

**Comment by matklad** [comment]
Ok, so let's r+ this!

bors r+

**Pull request merged**